File: Project2/Solutions/cpp/sudoku.cpp
#include "sudoku.h"


using std::cout;
using std::endl;
using std::queue;


/*
*   Sudoku Constructor
*/
Sudoku::Sudoku() {
    board.assign(81, 0);
}


/*
*   solve: High Level Function that calls both AC-3 and the solver
*/
bool Sudoku::solve() {
    CSP csp;

    // 1) Initialize the CSP problem for sudoku
    initCSP(&csp);

    // 2) Run AC-3 to reduce the domains for each variable
    if (!AC_3(&csp)) {
        cout << "Not consistent!" << endl;
        return false;
    }

    // 3) Solve the rest using search (assigning values we're sure of first)
    for (unsigned i = 0; i < 81; i++) {
        if (csp.D[i].size() == 1) {
            board[i] = csp.D[i][0];
        }
    }
    return search(&csp, 0);
}

void Sudoku::initCSP(CSP* csp) {
    // D (Domain)
    for (int i=0; i < 81; i++) {
        vector<int> domain;
        if (board[i] == 0) {
            for (int j = 1; j < 10; j++) {
                domain.push_back(j);
            }
        } else {
            domain.push_back(board[i]);
        }
        csp->D.push_back(domain);
    }

    // C (Constraints)
    for (int i=0; i < 81; i++) {
        vector<Arc> arcVec = arcConstraints(i);
        csp->C.push_back(arcVec);
    }
}


/*
*   AC_3 Algorithm
*
*   Returns false if an inconsistency is found, true otherwise
*/
bool Sudoku::AC_3(CSP* csp) {
    // generate all of the arcs in the csp
    queue<Arc> arcQ;
    for (unsigned i = 0; i < (csp->C).size(); i++) {
        for (unsigned j = 0; j < (csp->C)[i].size(); j++) {
            arcQ.push((csp->C)[i][j]);
        }
    }

    while (!arcQ.empty()) {
        Arc nextArc = arcQ.front(); arcQ.pop();
        if (revise(csp, nextArc)) {
            if ((csp->D)[nextArc.i].empty()) {
                return false;
            }
            vector<int> neighVec = neighbors(nextArc.i);
            for (unsigned i = 0; i < neighVec.size(); i++) {
                Arc arc(neighVec[i], nextArc.i);
                arcQ.push(arc);
            }
        }
    }

    return true;
}


/*
*   revise
*
*   Returns true IFF we revise the domain of arc.i
*/
bool Sudoku::revise(CSP* csp, Arc arc) {
    bool revised = false;
    unsigned maxI = (csp->D)[arc.i].size();

    for (unsigned i = 0; i < maxI; i++) {
        int x = (csp->D)[arc.i][i];
        bool sat = false;
        for (unsigned j = 0; j < (csp->D)[arc.j].size(); j++) {
            int y = (csp->D)[arc.j][j];
            if (constrSat(csp, x, y, arc.i, arc.j)) {
                sat = true;
                break;
            }
        }
        if (sat == false) {
            (csp->D)[arc.i].erase ( (csp->D)[arc.i].begin() + i );
            maxI--;
            revised = true;
        }
    }

    return revised;
}


/*
*   constrSat
*
*   Returns true if location i == x, location j == y satisfies constraints
*/
bool Sudoku::constrSat(CSP* csp, int x, int y, int i, int j) {
    if (x != y) return true;

    vector<Arc> constraints = (csp->C)[i];

    for (unsigned i = 0; i < constraints.size(); i++) {
        if (constraints[i].j == j) {
            return false;
        }
    }

    return true;
}

/*
*   Arc Constraints
*/
vector<Arc> Sudoku::arcConstraints(int ind) {
    vector<Arc> arcVec;
    vector<int> neighVec = neighbors(ind);

    for (unsigned i = 0; i < neighVec.size(); i++) {
        Arc arc(ind, neighVec[i]);
        arcVec.push_back(arc);
    }

    return arcVec;
}


/*
*   neighbors
*/
vector<int> Sudoku::neighbors(int ind) {
    vector<int> neighVec;

    int row, col;
    ind2rc(ind, &row, &col);

    // Add row neighbors
    int rInd = index(row, 0);
    for (int i = rInd; i < (rInd+9); i++) {
        if (i != ind) {
            neighVec.push_back(i);
        }
    }

    // Add col neighbors
    int cInd = index(0, col);
    for (int i = cInd; i < (cInd + 81); i = i + 9) {
        if (i != ind) {
            neighVec.push_back(i);
        }
    }

    // Add box neighbors (using bInd as our accessor)
    int bInd = index(row - (row % 3), col - (col % 3) );
    for (int i = 0; i < 9; i++) {
        if ( (i == 3) || (i == 6) ) {
            bInd = bInd + 6;
        }
        if (bInd != ind) {
            neighVec.push_back(bInd);
        }
        bInd++;
    }

    return neighVec;
}

/*
*   Backtracking search
*/
bool Sudoku::search(CSP* csp, int ind) {
    for (int i=ind; i < 81; i++) {
        if (board[i] != 0) continue;

        // Try every variable in the domain
        for (unsigned j = 0; j < ( (csp->D)[i].size() - 1); j++) {
            board[i] = (csp->D)[i][j];
            if (!validPlacement(i)) {
                continue;
            }
            if (search(csp, i+1)) {
                return true;
            }
        }
        board[i] = (csp->D)[i].back();
        if (validPlacement(i)) {
            if (search(csp, i+1)) {
                return true;
            }
        }
        board[i] = 0;
        return false;
    }

    return validBoard();
}

//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                         Validity Checkers                                //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////


/*
*   Returns true if no constraints are violated
*
*   Checks every constraint
*
*/
bool Sudoku::validBoard() {
    // Check Row Constraints
    for (int i=0; i < 9; i++) {
        if (!validRow(i)) {
            return false;
        }
    }

    // Check Col Constraints
    for (int j=0; j < 9; j++) {
        if (!validCol(j)) {
            return false;
        }
    }

    // Check Box Constraints
    for (int ii = 0; ii < 3; ii++) {
        for (int jj = 0; jj < 3; jj++) {
            if (!validBox(ii, jj)) {
                return false;
            }
        }
    }

    // otherwise return true
    return true;
}

/*
*   validPlacement
*
*   Returns true if value i doesn't violate any binary constraints
*/
bool Sudoku::validPlacement(int i) {
    int row, col;
    ind2rc(i, &row, &col);

    return (validRow(row) &&
            validCol(col) &&
            validBox(row/3, col/3));
}


/*
*   Verifies row is valid
*/
bool Sudoku::validRow(int i) {
    for (int j=0; j < 8; j++) {
        if (board[index(i, j)] == 0) continue;
        for (int k=j+1; k < 9; k++) {
            if (board[index(i, j)] == board[index(i, k)]) {
                return false;
            }
        }
    }
    return true;
}


/*
*   Verifies col is valid
*/
bool Sudoku::validCol(int j) {
    for (int i=0; i < 8; i++) {
        if (board[index(i, j)] == 0) continue;
        for (int k=i+1; k < 9; k++) {
            if (board[index(i, j)] == board[index(k, j)]) {
                return false;
            }
        }
    }
    return true;
}


/*
*   Verifies box is valid
*/
bool Sudoku::validBox(int ii, int jj) {
    for (int i=ii*3; i < (ii*3+3); i++) {
        for (int j=jj*3; j < (jj*3+3); j++) {
            if (board[index(i, j)] == 0) continue;
            for (int ki = i; ki < (ii*3+3); ki++) {
                for (int kj = jj*3; kj < (jj*3+3); kj++) {
                    if ( (ki == i) && (kj == j) ) continue;
                    if (board[index(i, j)] == board[index(ki, kj)]) {
                        return false;
                    }
                }
            }
        }
    }
    return true;
}


//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                         Helper  Functions                                //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////


/*
*   converts row i, col j, to vector index
*/
int Sudoku::index(int i, int j) {
    return (i*9 + j);
}


/*
*   converts vector index to row i, col j
*/
void Sudoku::ind2rc(int ind, int* row, int* col) {
    int row_, col_;

    row_ = 0;
    while (ind >= 9) {
        ind = ind - 9;
        row_++;
    }
    col_ = ind;

    (*row) = row_;
    (*col) = col_;
}


/*
*   Reads in "suinput.csv" and populates board
*
*   Returns 0 on success, 1 on failure
*/
int Sudoku::readInputFile() {
    std::ifstream inFile("suinput.csv");
    // std::ifstream inFile("suinput_worlds_hardest.csv");
    std::string line, val;

    // Read in the data
    if (inFile.is_open()) {
        for (int i=0; i < 9; i++) {
            if (getline(inFile, line)) {
                for (int j=0; j < 9; j++) {
                    val = line[j*2];
                    board[index(i, j)] = std::stoi(val);
                }
            } else {
                cout << "error reading suinput.csv" << endl;
                return 1;
            }
        }
        inFile.close();
        return 0;
    } else {
        cout << "Unable to open file" << endl;
        return 1;
    }
}


/*
*   Writes board to "suoutput.csv"
*/
void Sudoku::writeOutputFile() {
    std::ofstream outFile("suoutput.csv");

    for (int i=0; i < 9; i++) {
        for (int j=0; j < 9; j++) {
            outFile << board[index(i, j)];
            if (j != 8) outFile << ",";
        }
        if (i != 8) outFile << endl;
    }
    outFile.close();
}

/*
*   Writes board to "suoutput.csv"
*/
void Sudoku::writeOutputFileFail() {
    std::ofstream outFile("suoutput.csv");

    for (int i=0; i < 9; i++) {
        for (int j=0; j < 9; j++) {
            outFile << 0;
            if (j != 8) outFile << ",";
        }
        if (i != 8) outFile << endl;
    }
    outFile.close();
}

/*
*   writeToTerminal
*
*   Writes board to terminal
*/
void Sudoku::writeToTerminal() {
    for (int i=0; i < 9; i++) {
        for (int j=0; j < 9; j++) {
            cout << board[index(i, j)];
            if (j != 8) cout << ",";
        }
        cout << endl;
    }
    cout << endl;
}

File: Project #2/Solutions/python/sudoku.py
""" Part II - Problem 2 (Sudoku) """

import csv
from collections import deque

def revise(csp, arc):
    """ Revise function from AC-3 Algorithm """
    revised = False
    row_i, col_i = index_to_rc(arc[0])
    row_j, col_j = index_to_rc(arc[1])
    for x in csp.domains[row_i][col_i]:
        sat = False
        for y in csp.domains[row_j][col_j]:
            if x != y:
                sat = True
                break
        if not sat:
            csp.domains[row_i][col_i].remove(x)
            revised = True
    return revised


def ac_3(csp):
    """ AC-3 Algorithm - Return false if inconsistency found, true otherwise """
    # 1) Grab all of the constraint arcs
    arc_queue = deque()
    for i in range(0, 9):
        for j in range(0, 9):
            for arc in csp.constraints[i][j]:
                arc_queue.append(arc)

    # 2) Check each arc for consistency
    while len(arc_queue) > 0:
        next_arc = arc_queue.popleft()
        if revise(csp, next_arc):
            row, col = index_to_rc(next_arc[0])
            if len(csp.domains[row][col]) == 0:
                return False
            neighbor_idxs = neighbors(row, col)
            for neighbor_idx in neighbor_idxs:
                arc = (neighbor_idx, next_arc[0])
                arc_queue.append(arc)

    return True


def rc_to_index(i, j):
    """ Returns the linear index from the row and column """
    return i*9 + j

def index_to_rc(idx):
    """ Returns the row and column from the linear index """
    row = int(idx / 9)
    col = int(idx) % 9
    return row, col

def neighbors(row, col):
    """ Returns the constraint graph neighbors for index idx """
    neighbor_idxs = []

    # Add row neighbors
    for j in range(0, 9):
        if j != col:
            neighbor_idxs.append(rc_to_index(row, j))

    # Add col neighbors
    for i in range(0, 9):
        if i != row:
            neighbor_idxs.append(rc_to_index(i, col))

    # Add box neighbors
    box_row = int(row / 3) * 3
    box_col = int(col / 3) * 3
    for i in range(box_row, box_row + 3):
        for j in range(box_col, box_col + 3):
            if i != row and j != col:
                neighbor_idxs.append(rc_to_index(i, j))

    return  neighbor_idxs


def arc_constraints(row, col):
    """ Generate Arc Constraints for this row and col """
    arcs = []
    neighbor_idxs = neighbors(row, col)
    for neighbor_idx in neighbor_idxs:
        arc = (rc_to_index(row, col), neighbor_idx)
        arcs.append(arc)
    return arcs


class CSP:
    """ Constraint Satisfaction Problem """
    def __init__(self, board):
        self.domains = []
        self.constraints = []
        self.sudoku_init_domains(board)
        self.sudoku_init_constraints()

    def sudoku_init_domains(self, board):
        """ Initialize domains to 1-9 for every cell """
        for i in range(0, 9):
            row_domains = []
            for j in range(0, 9):
                domain = []
                if board[i][j] == 0:
                    for k in range(1, 10):
                        domain.append(k)
                else:
                    domain.append(board[i][j])
                row_domains.append(domain)
            self.domains.append(row_domains)

    def sudoku_init_constraints(self):
        """ Initialize constraints for the sudoku board """
        for i in range(0, 9):
            row_constraints = []
            for j in range(0, 9):
                arcs = arc_constraints(i, j)
                row_constraints.append(arcs)
            self.constraints.append(row_constraints)


class Sudoku:
    """ Sudoku Puzzle Class """
    def __init__(self):
        self.board = []
        self.read_input_file()
        self.csp = CSP(self.board)

    def read_input_file(self, fname="suinput.csv"):
        """ Read sudoku board puzzle from input file """
        with open(fname, 'r') as infile:
            csvreader = csv.reader(infile)
            lines = []
            for line in csvreader:
                lines.append(line)
            for i in range(0, 9):
                row_list = []
                for j in range(0, 9):
                    row_list.append(int(lines[i][j]))
                self.board.append(row_list)

    def write_output_file(self, fname="suoutput.csv"):
        """ Write Sudoku Board solution to output file """
        # 1) Open the file for writing
        f = open(fname, "w")

        # 2) Construct the output string
        out_str = ""
        for i in range(0, 9):
            out_str += str(self.board[i][0])
            for j in range(1, 9):
                out_str += "," + str(self.board[i][j])
            if i < 8:
                out_str += "\n"

        # 3) Write to the output file and then close it
        f.write(out_str)
        f.close()

    def write_output_file_fail(self, fname="suoutput.csv"):
        """ Write Failure Sudoku Board to output file """
        # 1) Open the file for writing
        f = open(fname, "w")

        # 2) Construct the output string
        out_str = ""
        for i in range(0, 9):
            out_str += str(0)
            for j in range(1, 9):
                out_str += "," + str(0)
            if i < 8:
                out_str += "\n"

        # 3) Write to the output file and then close it
        f.write(out_str)
        f.close()

    def solve(self):
        """ Solve the Sudoku Puzzle """
        # 1) Run AC-3
        if not ac_3(self.csp):
            print("AC-3 returned Inconsistent!")
            return False

        # 2) Assign Values we're sure of
        for i in range(0, 9):
            for j in range(0, 9):
                if len(self.csp.domains[i][j]) == 1:
                    self.board[i][j] = self.csp.domains[i][j][0]

        # 3) Solve the rest using search on the limited domains
        return self.search(0)


    def search(self, idx):
        """ Backtracking search """
        # Loop over every space that hasn't been filled yet
        row, col = index_to_rc(idx)
        for i in range(row, 9):
            for j in range(col, 9):
                if self.board[i][j] != 0:
                    continue

                # Try every value in its domain
                curr_idx = rc_to_index(i, j)
                for val in self.csp.domains[i][j]:
                    self.board[i][j] = val
                    if not self.valid_placement(i, j):
                        continue

                    if self.search(curr_idx+1):
                        return True

                # And if none of them work, set back to 0 and return failure
                self.board[i][j] = 0
                return False


        return self.valid_board()

    def valid_board(self):
        """ Constraint checker on whole board """
        # 1) Check Row Constraints
        for i in range(0, 9):
            if not self.valid_row(i):
                return False

        # 2) Check Col Constraints
        for j in range(0, 9):
            if not self.valid_col(j):
                return False

        # 3) Check Box Constraints
        for i in range(0, 3):
            for j in range(0, 3):
                if not self.valid_box(i*3, j*3):
                    return False

        # Otherwise return true
        return True

    def valid_placement(self, row, col):
        """ Constraint checker for a specific cell """
        return self.valid_row(row) and \
               self.valid_col(col) and \
               self.valid_box(int(row/3)*3, int(col/3)*3)


    def valid_row(self, row):
        """ Constraint checker on a specific row """
        for j in range(0, 8):
            if self.board[row][j] == 0:
                continue
            for k in range(j+1, 9):
                if self.board[row][j] == self.board[row][k]:
                    return False
        return True

    def valid_col(self, col):
        """ Constraint checker on a specific col """
        for i in range(0, 8):
            if self.board[i][col] == 0:
                continue
            for k in range(i+1, 9):
                if self.board[i][col] == self.board[k][col]:
                    return False
        return True

    def valid_box(self, row, col):
        """ Constraint checker on a specific box """
        values = []
        for i in range(row, row + 3):
            for j in range(col, col + 3):
                values.append(self.board[i][j])
        for k in range(1, 10):
            if values.count(k) > 1:
                return False
        return True


def main():
    """ main function to call sudoku code """
    # 1) Read in the sudoku board
    my_sudoku = Sudoku()

    # 2) Solve the sudoku board
    valid_solution = my_sudoku.solve()

    # 3) Write the solution
    if valid_solution:
        print("Success!")
        my_sudoku.write_output_file()
    else:
        print("Failure!")
        my_sudoku.write_output_file_fail()


def test_input_output():
    """ Test Code for input output """
    my_sudoku = Sudoku()
    my_sudoku.write_output_file()


if __name__ == "__main__":
    main()
    # test_input_output()
